lib/Interface/cpp_interface.cpp

```diff
@@ -613,4 +640,47 @@ void CNFClearMemory()
 {
   Cnf_ClearMemory();
 }
+
+Cpp_interface::SolverFrame::SolverFrame(
+    std::unordered_map<std::string, Function>* global_function_context)
+    : _global_function_context(global_function_context)
+{
+}
+
+// When we destroy a solver frame, we need to make sure that all of the scoped
+// functions in the global function context are also correctly removed.
+//
+// This ensures that the reference counting for any symbols used in the
+// function declarations are correctly decremented.
+Cpp_interface::SolverFrame::~SolverFrame()
+{
+  // Iterate on the function names in our current scope
+  for (vector<std::string>::const_iterator scoped_function_name =
+           getFunctions().begin();
+       scoped_function_name != getFunctions().end(); ++scoped_function_name)
+  {
```

Comment:
Feel free to use modern style if you'd like:
``
for (const auto& scoped_function_name : getFunctions())
``